### PR TITLE
[ONNX] Opset 11 updates

### PR DESCRIPTION
**Pull request by neginraoof**
This PR contains:
1- pad updates for opset11 symbolic
2- Updated avg_pool for opset11
3- TopK updates for opset 11

### CODE REVIEW

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -226,17 +226,18 @@ def _sort_helper(g, input, dim, decending=True, out=None):
 def _topk_helper(g, input, k, dim, largest=True, sorted=False, out=None):
     if out is not None:
         _unimplemented("TopK", "Out parameter is not supported")
+    if not largest:
+        _unimplemented("TopK", "Ascending is not supported")
+    k = _maybe_get_const(k, 'i')
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py]: Why move these 3 lines outside of the if branch? These are not required for the else branch

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -226,17 +226,18 @@ def _sort_helper(g, input, dim, decending=True, out=None):
...
+        k = g.op("Constant", value_t=torch.tensor(k, dtype=torch.int64))
     if _export_onnx_opset_version <= 10:
-        if not largest:
-            _unimplemented("TopK", "Ascending is not supported")
-            return g.op("TopK", input, k_i=k, axis_i=dim, outputs=2)
-        k = _maybe_get_const(k, 'i')
-        if not _is_value(k):
-            k = g.op("Constant", value_t=torch.tensor(k, dtype=torch.int64))
         from torch.onnx.symbolic_opset9 import unsqueeze
         k = unsqueeze(g, k, 0)
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py]: is `k` required to be 1-D tensor for ONNX::TopK? Reshaping to shape [1] is safer because unsqueeze assumes previous `k` is always scalar.

File: torch/onnx/symbolic_opset11.py [hunk 3]
```diff
@@ -180,6 +180,60 @@ def round(g, self):
     return g.op("Round", self)
 
 
+# Generate paddings in ONNX order based on pad in pytorch.
+# Arguments:
+#     dim: the dimension of the tensor.
+#     pad: the paddings in pytorch.
+#          The order is dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, ...
+def _prepare_onnx_paddings(g, dim, pad):
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: The logic is fine here. This supports paddings of arbitrary or unknown shape. A few suggestions.

1. Please add more comments with examples illustrating what each step is doing.
2. `Pad` can be replaced with `Concat`. This should make the code cleaner as it would get rid of the Casting.
3. Fuse the last step `Reshape` to [1, -1], and `Squeeze` [0] to `Reshape` to [-1].
4. Use `torch.onnx.symbolic_opset9.size` to get pad_len.

**Review by BowenBao — CHANGES_REQUESTED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -1371,6 +1357,35 @@ def forward(self, x):
         x = torch.tensor([0.9920, -1.0362, -1.5000, 3.5000], requires_grad=True)
         self.run_test(Round(), x)
 
+    # TODO: enable this test once ORT supports more input types
+    @skipIfUnsupportedOpsetVersion([11])
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: value attribute is double in test_constant_pad. This is a missing type in ORT.

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -226,17 +226,18 @@ def _sort_helper(g, input, dim, decending=True, out=None):
 def _topk_helper(g, input, k, dim, largest=True, sorted=False, out=None):
     if out is not None:
         _unimplemented("TopK", "Out parameter is not supported")
+    if not largest:
+        _unimplemented("TopK", "Ascending is not supported")
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py]: Fixed

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py]: Fixed

File: torch/onnx/symbolic_helper.py (hunk 2 above)
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py]: Yes it is. I updated to reshape. Thanks.

File: torch/onnx/symbolic_opset11.py (hunk 3 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset11.py]: Thanks for the suggestion. As we talked, I updated this Concat + ConstantOfShape.

**Comment by neginraoof** [comment]
Looking into fixing test_onnx_opset test_topk

File: test/test_utils.py [hunk 6]
```diff
@@ -452,12 +451,6 @@ def test_smoke(self):
 
 
 class TestONNXUtils(TestCase):
-    def test_prepare_onnx_paddings(self):
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/test_utils.py]: Sounds good. I was thinking that the functionality will be tested in ORT tests, but I've added this back for now.

File: torch/onnx/symbolic_helper.py [hunk 7]
```diff
@@ -228,15 +228,12 @@ def _sort_helper(g, input, dim, decending=True, out=None):
 def _topk_helper(g, input, k, dim, largest=True, sorted=False, out=None):
     if out is not None:
         _unimplemented("TopK", "Out parameter is not supported")
+    if not _is_value(k):
+        k = g.op("Constant", value_t=torch.tensor(k, dtype=torch.int64))
+    k = g.op("Reshape", k, g.op("Constant", value_t=torch.tensor([1])))
```
Comment (BowenBao) [inline torch/onnx/symbolic_helper.py]: nit: `Reshape` can be put in an else branch. In the if branch construct the `Constant` with torch.tensor([k], dtype=torch.int64).

File: torch/onnx/symbolic_opset11.py [hunk 8]
```diff
@@ -180,6 +180,62 @@ def round(g, self):
...
+    # dim_0_begin, dim_1_begin, ... , dim_0_end, ..., dim_n_end.
+    # n is the dimension of input.
+    # Assume zero-dimensions in the beginning, pad the "pad" sequence with zeros in the beginning
+    pad_len = torch.onnx.symbolic_opset9.size(g, pad, g.op("Constant", value_t=torch.tensor([0])))
+    # Set extension = [0] * (dim * 2 - len(pad))
+    extension = g.op("Sub", g.op("Mul", g.op("Constant", value_t=torch.tensor(dim, dtype=torch.int64)),
+                     g.op("Constant", value_t=torch.tensor(2, dtype=torch.int64))), pad_len)
+    # Concat pad with extension: paddings = [dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, 0, 0, ... ]
+    paddings = g.op("Concat", pad, g.op("ConstantOfShape", extension, value_t=torch.tensor([0])), axis_i=0)
+    # ext_pad = g.op("Transpose", ext_pad, perm_i=[1, 0])
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: remove commented code

File: torch/onnx/symbolic_opset11.py [hunk 9]
```diff
@@ -180,6 +180,62 @@ def round(g, self):
...
+    # Reshape and reverse order and collate first beginnings and then ends
+    # paddings = [[..., 0, dim_n - 1_begin, dim_n_begin]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]:
```suggestion
    # paddings = [[..., 0, dim_n-1_begin, dim_n_begin],
```

File: torch/onnx/symbolic_opset11.py [hunk 10]
```diff
@@ -180,6 +180,62 @@ def round(g, self):
...
+    #               [..., 0, dim_n - 1_end, dim_n_end]]
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]:
```suggestion
    #               [..., 0, dim_n-1_end, dim_n_end]]
```

File: torch/onnx/symbolic_opset11.py [hunk 11]
```diff
@@ -180,6 +180,62 @@ def round(g, self):
     return g.op("Round", self)
 
 
+# Generate paddings in ONNX order based on pad in pytorch.
+# Arguments:
+#     dim: the dimension of the tensor.
+#     pad: the paddings in pytorch.
+#          The order is dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, ...
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]:
```suggestion
#          The order is dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, ..., dim_m_begin, dim_m_end,
#          where m is in range [0, n].
```
The original comment was unclear that PyTorch padding could possibly omit zero paddings.

File: torch/onnx/symbolic_opset11.py [hunk 12]
```diff
@@ -180,6 +180,62 @@ def round(g, self):
...
+    paddings = g.op("Concat", pad, g.op("ConstantOfShape", extension, value_t=torch.tensor([0])), axis_i=0)
+    # ext_pad = g.op("Transpose", ext_pad, perm_i=[1, 0])
+    # Reshape and reverse order and collate first beginnings and then ends
+    # paddings = [[..., 0, dim_n - 1_begin, dim_n_begin]
+    #               [..., 0, dim_n - 1_end, dim_n_end]]
+    # Reshape back to 1-D paddings = [..., 0, dim_n - 1_begin, dim_n_begin, ..., 0, dim_n - 1_end, dim_n_end]
+    paddings = g.op("Reshape", paddings, g.op("Constant", value_t=torch.tensor([-1, 2])))
+    paddings = g.op("Transpose", torch.onnx.symbolic_opset10.flip(g, paddings, [0]), perm_i=[1, 0])
+    paddings = g.op("Reshape", paddings, g.op("Constant", value_t=torch.tensor([-1])))
+    padding_c = g.op("Cast", paddings, to_i=sym_help.cast_pytorch_to_onnx['Long'])
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: Do we still need this cast?

Comment (neginraoof) [reply torch/onnx/symbolic_opset11.py]: Yes, this will be needed in any case cause pad should be long as per onnx spec

File: torch/onnx/symbolic_helper.py (hunk 7 above)
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py]: That won't work for opset 10, if k is value. I updated this with a minor change.

**Review by spandantiwari — APPROVED** [review]
Overall LGTM. Added one comment that needs attention.

File: torch/onnx/symbolic_helper.py [hunk 13]
```diff
@@ -368,6 +366,20 @@ def _index_fill_reshape_helper(g, self, dim, index):
...
 
+
+def _avgpool_count_include_pad_helper(g, input, padding):
+    if _export_onnx_opset_version <= 10:
+        input = g.op("Pad", input,
+                     pads_i=((0,) * 2 + padding) * 2,
+                     mode_s='constant',
+                     value_f=0.)
+    else:
+        input = g.op("Pad", input,
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/onnx/symbolic_helper.py]: Thanks Spandan. I agree with you about scenarios like this with different branches for opset versions. I think I can keep this helper for opsets 9 and 10, and implement this for opset 11 in its own file.

**Comment by neginraoof** [comment]
cc @houseroad please review

**Comment by neginraoof** [comment]
cc @houseroad please review

File: torch/onnx/symbolic_opset11.py [hunk 14]
```diff
@@ -180,6 +203,62 @@ def round(g, self):
...
+    # The desired order of paddings is
+    # dim_0_begin, dim_1_begin, ... , dim_0_end, ..., dim_n_end.
+    # n is the dimension of input.
+    # Assume zero-dimensions in the beginning, pad the "pad" sequence with zeros in the beginning
+    pad_len = torch.onnx.symbolic_opset9.size(g, pad, g.op("Constant", value_t=torch.tensor([0])))
+    # Set extension = [0] * (dim * 2 - len(pad))
+    extension = g.op("Sub", g.op("Mul", g.op("Constant", value_t=torch.tensor(dim, dtype=torch.int64)),
+                     g.op("Constant", value_t=torch.tensor(2, dtype=torch.int64))), pad_len)
+    # Concat pad with extension: paddings = [dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, 0, 0, ... ]
+    paddings = g.op("Concat", pad, g.op("ConstantOfShape", extension, value_t=torch.tensor([0])), axis_i=0)
```
Comment (dashesy) [inline torch/onnx/symbolic_opset11.py]: I get this error from this line:

> Type parameter (T) bound to different types (tensor(int32) and tensor(int64) in node (__Concat_213).

Comment (neginraoof) [reply torch/onnx/symbolic_opset11.py]: ONNX only supports int64 for now. I added the cast, but you cannot have int pads in ONNX for now.

**Comment by neginraoof** [comment]
@houseroad Could you please let me know if there are any related failures?

**Review by houseroad — APPROVED** [review]
LGTM
